linkerd/app/core/src/control.rs

```diff
@@ -145,174 +151,74 @@ pub mod add_origin {
     }
 }
 
-/// Resolves the controller's `addr` once before building a client.
 pub mod resolve {
-    use super::{client, ControlAddr};
-    use crate::svc;
-    use futures::{ready, TryFuture};
-    use linkerd2_addr::Addr;
-    use linkerd2_dns as dns;
-    use pin_project::pin_project;
-    use std::future::Future;
+    use super::client::Target;
+    use crate::{
+        dns,
+        proxy::{discover, dns_resolve::DnsResolve, resolve::map_endpoint},
+        svc,
+    };
     use std::net::SocketAddr;
-    use std::pin::Pin;
-    use std::task::{Context, Poll};
-    use std::{error, fmt};
 
-    #[derive(Clone, Debug)]
-    pub struct Layer {
+    pub fn layer<M>(
         dns: dns::Resolver,
+    ) -> impl svc::Layer<
+        M,
+        Service = discover::MakeEndpoint<
+            discover::FromResolve<map_endpoint::Resolve<IntoTarget, DnsResolve>, Target>,
+            M,
```

Comment:
not a blocker but could we maybe make this thing a type alias or something?

Reply:
it's only used the once, isn't it? so, yes, but it wouldn't really reduce anything